go/consensus/cometbft/apps/vault/transactions.go

```diff
@@ -0,0 +1,213 @@
+	ctx = ctx.NewTransaction()
+	defer ctx.Close()
+
+	// Determine caller nonce.
+	stakeState := stakingState.NewMutableState(ctx.State())
+	callerAcct, err := stakeState.Account(ctx, ctx.CallerAddress())
+	if err != nil {
+		return err
+	}
+
+	// Create a new vault.
+	newVault := &vault.Vault{
+		CreatedBy:        ctx.CallerAddress(),
+		CreatedNonce:     callerAcct.General.Nonce,
+		AdminAuthority:   args.AdminAuthority,
+		SuspendAuthority: args.SuspendAuthority,
+	}
+	if err = state.CreateVault(ctx, newVault); err != nil {
+		return err
+	}
+
+	ctx.Commit()
+
+	return nil
+}
+
+func (app *vaultApplication) authorizeAction(ctx *api.Context, args *vault.AuthorizeAction) error {
+	state := vaultState.NewMutableState(ctx.State())
+	params, err := state.ConsensusParameters(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch consensus parameters: %w", err)
+	}
+
+	if err = args.Validate(params); err != nil {
+		return fmt.Errorf("%w: %w", vault.ErrInvalidArgument, err)
+	}
+
+	// Ensure vault exists.
+	vlt, err := state.Vault(ctx, args.Vault)
+	if err != nil {
+		return err
+	}
+
+	// Validate action nonce. Currently queuing multiple future actions is not allowed.
+	if args.Nonce != vlt.Nonce {
+		return vault.ErrInvalidNonce
+	}
+
+	// Validate whether the caller is authorized to submit an action.
+	if !args.Action.IsAuthorized(vlt, ctx.CallerAddress()) {
+		return vault.ErrForbidden
+	}
+
+	if ctx.IsCheckOnly() {
+		return nil
+	}
+
+	// Charge gas for this transaction.
+	if err = ctx.Gas().UseGas(1, vault.GasOpAuthorizeAction, params.GasCosts); err != nil {
+		return err
+	}
+
+	// Return early for simulation as we only need gas accounting.
+	if ctx.IsSimulation() {
+		return nil
+	}
+
+	// Start a new transaction and rollback in case we fail.
+	ctx = ctx.NewTransaction()
+	defer ctx.Close()
+
+	var pendingAction *vault.PendingAction
+	// Check if there is an existing pending action for this nonce. In this case, we will be
+	// updating the action.
+	pendingAction, err = state.PendingAction(ctx, args.Vault, args.Nonce)
+	switch err {
+	case nil:
+		// Pending action already exists.
+		if !pendingAction.Action.IsAuthorized(vlt, ctx.CallerAddress()) {
+			return vault.ErrForbidden
+		}
+
+		// If this is not a cancellation, ensure that the action is the same so that the authorizer
+		// really signed the correct action.
+		if args.Action != nil && !pendingAction.Action.Equal(args.Action) {
+			return vault.ErrInvalidArgument
+		}
+	case vault.ErrNoSuchAction:
+		// In case there is no existing action, this must not be a cancellation.
+		if args.Action == nil {
+			return vault.ErrInvalidArgument
+		}
+
+		// Create a new pending action.
+		pendingAction = &vault.PendingAction{
+			Nonce:     args.Nonce,
+			CreatedAt: ctx.BlockHeight() + 1, // Current height is ctx.BlockHeight() + 1.
+			Action:    *args.Action,
+		}
+	default:
+		return err
+	}
+
+	var removeAction bool
+	switch args.Action {
+	case nil:
+		// Cancel existing action.
```

Comment:
Does this mean that a single authority address can cancel a pending action, even if the authority threshold is >1? 

Could a single compromised/malicious authority address then prevent updates for a specific vault by constantly submitting cancel action transactions for any pending action?

Reply:
Yes. The idea is that this is a safety feature so malicious/errorneous transactions are easier to abort. This is modeled after other existing multi-sig implementations, but it is indeed an availability tradeoff in case of malicious actors.

Reply:
Ok. Could a simple improvement be to at least require an Admin authority for cancelation? So that suspend authority (and any other added authority in future) cannot prevent Admin actions?